research/06-require-ana.tex

```diff
@@ -1,9 +1,184 @@
+enables the client to track the progress and finally it allows validation of the
+deliverable. These requirements are listed in this section.
+
+\subsection{Design goals}
+\label{ssec:goals}
+
+\todo{The line between design goals and requirements is thin. Perhaps some of
+the design goals need to be moved down.}
+
+The client has expressed some high-level requirements, which are listed in this
+subsection as design goals in order of prioroty. The design goals serve as an
+important guideline when defining and implement the invidual requirements. As
+such, they can be considered the bounds within which all requirements must fit.
+
+\paragraph{Language-agnostic} A REPL should be generated for each language
+defined with Spoofax. This means that no assumptions can be made about any
+language constructs or syntax.
+
+\paragraph{Autogeneration} The REPLs should be automatically generated, just as
+all of Spoofax's services. This means that it should be integrated into
+Spoofax's build system and not require additional steps from the user.
+
+\paragraph{Maintainability} Spoofax is an already existing, open source project
+managed by several people. When the product is delivered, it will have to be
+maintained by others. Therefore, it is important that the code is maintainable.
+
+\paragraph{Performance} Spoofax's developers already focus quite a bit on
+performance: both the generation and the use of the services are performant.
+This should be no different for the generation and use of the REPL.
+
+\paragraph{Modify Spoofax's existing codebase as little as possible} The product
+should be an extension to Spoofax, which means the changes made to the existing
+Spoofax codebase should be as small as possible. Preferably, the REPL service
+generator should be a standalone module.
+
+\subsection{Requirements}
+\label{ssec:requirements}
+
+Under the guidance of the design goals listed in the previous subsection, the
+requirements compiled from the feature matrix discussed in \cref{sec:repl} and
+meetings with the client are discussed below, under the MoSCoW method.
+
+\subsubsection{Must have}
+
+Requirements listed under ``must haves'' are of critical importance to the
+usability and success of the deliverable. Without these, the product is not in a
+workable state and is not likely to be accepted by the client. \textit{Must} can
+also be considered an acronym for the Minimum Usable SubseT.
+
+\paragraph{Interactive shell} Per the definition of a REPL given in
+\cref{sec:repl}, the generated REPLs have to be interactive. It should evaluate
+single statements and expressions typed in by the user and print the results
+back to them.
+
+\paragraph{Generated from language definition} Every service generated by
+Spoofax is made from the language definitions. It is evident that the REPL
+should be generated from the same definitions if it is to fit within Spoofax.
+
+\todo{The following requirement is also a design goal.}
```

Comment:
See line 20, the language-agnostic design goal. It's just that currently it is listed as a requirement, whilst perhaps it should be a design goal.